lib/conrad/emitter_queue.rb

```diff
@@ -0,0 +1,52 @@
+module Conrad
+  class EmitterQueue
+    include Singleton
+
+    attr_reader :background
+    attr_accessor :logger
+
+    def initialize
+      Thread.abort_on_exception = true
+      @queue = Queue.new
+      @logger ||= Logger.new(STDOUT)
+    end
+
+    def background=(value)
+      @background = value
+      start_thread if value
+    end
+
+    def enqueue
+      @queue.push -> () { yield }
+      
+      # if it's backgounded we can break out of here, as the background
+      #   queue will pick it up. otherwise, we need to explicitly process it
+      process! unless background
+    end
+
+    private
+
+    def process!
+      until @queue.empty? do
+        emit!
+      end
+    end
+
+    def start_thread
+      @thread ||= Thread.new do
+        loop do
+          unless @queue.empty?
```

Comment:
Why not do the same `until @queue.empty?` syntax here? I'm not totally sure if there are semantic reasons or differences between them but it just seems strange to have two forms of what looks like the same logic.